x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/create_visualization.ts

```diff
@@ -0,0 +1,264 @@
+          }
+        }
+
+        // Step 2: Get or generate ES|QL query
+        let esqlQuery: string;
+
+        if (!esql) {
+          logger.debug('No ES|QL query or index provided, generating one');
+          const generateEsqlResult = await runner.runTool({
+            toolId: platformCoreTools.generateEsql,
+            toolParams: {
+              query: nlQuery,
+            },
+          });
+
+          // Extract the ES|QL query from the results
+          const queryResult = generateEsqlResult.results.find(
+            (result: ToolResult) => result.type === ToolResultType.query
+          );
+
+          if (
+            !queryResult ||
+            !queryResult.data ||
+            !('esql' in queryResult.data) ||
+            typeof queryResult.data.esql !== 'string'
+          ) {
+            throw new Error('Failed to generate ES|QL query');
+          }
+
+          esqlQuery = queryResult.data.esql;
+          logger.debug(`Generated ES|QL query: ${esqlQuery}`);
+        } else {
+          // Looks like an ES|QL query
+          esqlQuery = esql;
+          logger.debug('Using provided ES|QL query');
+        }
+
+        // Step 3: Generate visualization configuration with validation retry loop
+        const model = await modelProvider.getDefaultModel();
+        let validatedConfig: any | null = null;
+
+        const schema = parse(esqlMetricState.getSchema());
+
+        const systemPrompt = `You are a Kibana Lens visualization configuration expert. Generate a valid JSON configuration for a ${selectedChartType} visualization based on the provided schema and ES|QL query.
+
+Schema for ${selectedChartType}:
+${JSON.stringify(schema, null, 2)}
+
+${existingConfig ? `Existing configuration to modify: ${existingConfig}` : ''}
+
+IMPORTANT RULES:
+1. The response must be valid JSON only, no markdown or explanations
+2. The 'dataset' field must contain: { type: "esql", query: "<the provided ES|QL query>" }
+3. always use { operation: 'value', column: '<esql column name>', ...other options } for operations
+4. All field names must match those available in the ES|QL query result
+5. Make sure to follow schema definition strictly`;
+
+        logger.debug(`System prompt: ${systemPrompt}`);
+
+        // Build conversation messages array that will accumulate across retries
+        const conversationMessages: Array<{ role: string; content: string }> = [
+          {
+            role: 'system',
+            content: systemPrompt,
+          },
+          {
+            role: 'user',
+            content: `User query: ${nlQuery}
+              
+ES|QL query: ${esqlQuery}
+
+Generate the ${selectedChartType} visualization configuration in valid JSON format.`,
+          },
+        ];
+
+        for (let attempt = 1; attempt <= MAX_RETRY_ATTEMPTS; attempt++) {
+          logger.debug(
+            `Attempting to generate visualization configuration (attempt ${attempt}/${MAX_RETRY_ATTEMPTS})`
+          );
```

Comment:
Fine for a POC, but I'd rather use langgraph for flow control here, for consistency with our other "smart tool" flows and because it's just so much more readable for multi-step flows, especially when you start adding retry-on-failure or similar logic.

WDYT?

Reply:
yes, i wanted to keep this simple for PoC sake, but the final implementation should do things the same way we do elsewhere.